File: Codes/Kan Zheng/C++/linkedlist/328.Odd Even Linked List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* oddEvenList(ListNode* head) {
        if(head==NULL||head->next==NULL)
            return head;
        ListNode* l1=head;
        ListNode* l2=head->next;
        ListNode* Node1=l1;
        ListNode* Node2=l2;
        bool flag=1;
        for(;;)
        {
            if(Node1->next!=NULL&&Node2->next!=NULL)
            { 
                if(flag)
                {
                    Node1->next = Node2->next;
                    Node1 = Node1->next;
                }
                else
                {
                    Node2->next = Node1->next;
                    Node2 = Node2->next;
                }
                flag = 1- flag;
            }
            else
            {
                Node1->next = l2;
                Node2->next = NULL;
                break;
            }
        }
        return l1;
    }
};

File: Codes/Bingran Li/LinkList/328. Odd Even Linked List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode oddEvenList(ListNode head) {
        // solution 1: memory exceed??!!------have bug
     
        ListNode oddHead,evenHead;
        ListNode node1, node2;
        if(head==null || head.next==null)
            return head;
        oddHead = head;
        evenHead = head.next;
        node1=oddHead;
        node2=evenHead;
        while(node2!=null && node2.next!=null){ // miss first condition
            node1.next=node2.next;
            node1=node2.next;
           // if(node1.next==null)  -----> wrong
           //   break;
            node2.next=node1.next;
            node2=node1.next;
        }
        node1.next = evenHead;
        return head;

        // solution 2: 
        
        
    }
}